### PR TITLE
#5852 Neo Input and Neo Field

### PR DESCRIPTION
**Pull request by prachi00**
**Thank you for your contribution** to the [KodaDot - One Stop Shop for Polkadot NFTs](https://kodadot.xyz).

👇 __ Let's make a quick check before the contribution.

## PR Type

- [ ] Bugfix
- [ ] Feature
- [x] Refactoring

## Context

- [ ] Closes #
- [ ] Requires deployment <snek/rubick/worker>

#### Before submitting pull request, please make sure:

- [ ] My contribution builds **clean without any errors or warnings**
- [ ] I've merged recent default branch -- **main** and I've no conflicts
- [ ] I've tried to respect high code quality standards
- [ ] I've didn't break any original functionality

#### Optional

- [ ] I've tested it at </ksm/collection>
- [ ] I've tested PR on mobile
- [ ] I've written unit tests 🧪
- [ ] I've found edge cases

#### Had issue bounty label?

- [x] Fill up your KSM address: 
[Payout]https://kodadot.xyz/dot/transfer?target=13Qx65nLd6SwdtjrRyuoEtp9CKXhF651xdHBPaXcvhwKm4N1

#### Community participation

- [x] [Are you at KodaDot Discord?](https://discord.gg/35hzy2dXXh)

## Screenshot 📸

- [ ] My fix has changed **something** on UI; a screenshot is best to understand changes for others.
SS
Screen Shot 2023-05-08 at 12 33 33 PM
Screen Shot 2023-05-08 at 12 33 50 PM

Screen Shot 2023-05-08 at 12 34 09 PM
Screen Shot 2023-05-08 at 12 34 37 PM

### <samp>🤖 Generated by Copilot at e1b9ba3</samp>

No summary available (Limit exceeded: required to process 75846 tokens, but only 50000 are allowed per call)

### <samp>🤖 Generated by Copilot at 7b7ed90</samp>

> _We are the NeoFields, we rise from the ashes_
> _We replace the b-fields, we bring the new flashes_
> _We are the NeoFields, we rule the forms and grids_
> _We crush the b-fields, we are the Neo kids_

### CODE REVIEW

**Comment by vikiival** [comment]
@prachi00 can you provide screenshot pls?

File: libs/ui/src/components/NeoInput/NeoInput.scss
Comment (daiagi) [inline libs/ui/src/components/NeoInput/NeoInput.scss:1]: you forgot to import ```input.scss``` from oruga
you can override some of oruga's styling

have a look here:

https://github.com/kodadot/nft-gallery/blob/ebe85045843b59d0cc20aca0b8bcdc911ec6a79d/libs/ui/src/components/NeoInput/NeoInput.scss

File: libs/ui/src/components/NeoField/NeoField.scss
Comment (daiagi) [inline libs/ui/src/components/NeoField/NeoField.scss:1]: need to import utiltities from oruga and field component

File: libs/ui/src/components/NeoInput/NeoInput.scss
Comment (prachi00) [reply libs/ui/src/components/NeoInput/NeoInput.scss:1]: imported

**Comment by prachi00** [comment]
SS

> @prachi00 can you provide screenshot pls?

added some

**Comment by vikiival** [comment]
@prachi00 can you please resolve conflicts?

**Comment by prachi00** [comment]
> @prachi00 can you please resolve conflicts?

done

**Review by vikiival — COMMENTED** [review]
~

File: libs/ui/src/components/NeoField/NeoField.vue [hunk 1]
```diff
@@ -0,0 +1,23 @@
...
+  <o-field
+    :label="label"
+    :grouped="grouped"
+    :variant="variant"
+    class="neo-field">
+    <slot />
+  </o-field>
+</template>
+
+<script lang="ts" setup>
```
[the remark this answers is not in the document]
Comment (prachi00) [reply libs/ui/src/components/NeoField/NeoField.vue:11]: we do need some custom css though like 
Screen Shot 2023-05-15 at 8 56 25 PM

File: components/explore/ExploreGrid.vue [hunk 2]
```diff
@@ -19,13 +19,14 @@
...
       </p>
-    </b-field>
+    </NeoField>
   </div>
 </template>
 
 <script setup lang="ts">
 import { NeoButton } from '@kodadot1/brick'
 import { usePreferencesStore } from '@/stores/preferences'
+import { NeoField } from '@kodadot1/brick'
```
Comment (roiLeo) [inline components/explore/ExploreGrid.vue:29]:
```suggestion
import { NeoButton, NeoField } from '@kodadot1/brick'
import { usePreferencesStore } from '@/stores/preferences'
```

File: components/rmrk/Create/SimpleMint.vue [hunk 3]
```diff
@@ -142,14 +142,17 @@
...
+    <NeoField type="is-danger" :message="haveNoToSMessage">
       <b-switch v-model="hasToS" :rounded="false" data-cy="input-tos">
         {{ $t('termOfService.accept') }}
       </b-switch>
-    </b-field>
-    <b-field v-if="isLogIn" type="is-danger" :message="balanceNotEnoughMessage">
+    </NeoField>
+    <NeoField
+      v-if="isLogIn"
+      type="is-danger"
```
Comment (roiLeo) [inline components/rmrk/Create/SimpleMint.vue:154]: ?
```suggestion
      variant="danger"
```

**Review by roiLeo — COMMENTED** [review]
please could you check 
- `type="is-...` => `variant="..."`
- remove `message` prop on NeoField

**Comment by vikiival** [comment]
Agree with @roiLeo

**Comment by vikiival** [comment]
@kodadot/qa-guild is it possible to check this ? Seems quite a big change 😶

**Comment by prachi00** [comment]
> please could you check
> 
> * `type="is-...` => `variant="..."`
> * remove `message` prop on NeoField

why are we removing message props? it seems it is supported by oruga https://oruga.io/components/field.html#props
@roiLeo

**Review by roiLeo — APPROVED** [review]
> why are we removing message props? it seems it is supported by oruga https://oruga.io/components/field.html#props

Nevermind, since I didn't saw `message` prop in`NeoField` I thought it wasn't working 
Screenshot 2023-05-17 at 08-40-39 KodaDot Low fees and low carbon minting KodaDot - NFT Market Explorer


✅ lgtm

**Comment by vikiival** [comment]
@prachi00 can you resolve conflicts please?

File: components/rmrk/Create/ArweaveUploadSwitch.vue [hunk 4]
```diff
@@ -20,6 +20,10 @@ import { NeoField } from '@kodadot1/brick'
     NeoField,
   },
 })
+
+import { NeoSwitch } from '@kodadot1/brick'
+
+@Component({ components: { NeoSwitch } })
```
Comment (preschian) [inline components/rmrk/Create/ArweaveUploadSwitch.vue:26]: the build failed, `@Component` declared multiple times in here

**Comment by prachi00** [comment]
some test cases are failing after resolving conflicts, I'll fix them tomorrow

**Comment by VladoTheBoi** [comment]
@prachi00 is it possible to make a little change ?
I am not sure how it behaves rn but ideally have it like this for light and dark:
Screenshot 2023-05-18 at 12 33 16

**Comment by prachi00** [comment]
@preschian @roiLeo  test cases are failing for this part when updating to neoInput, but it seems like we dont even use this function, can we remove it? can someone confirm? 
Screen Shot 2023-05-18 at 7 41 05 PM

File: libs/ui/src/components/NeoInput/NeoInput.vue [hunk 5]
```diff
@@ -14,7 +23,17 @@ import { OInput } from '@oruga-ui/oruga'
...
+  type?: string
   placeholder?: string
+  maxlength?: number
+  min?: number
+  max?: number
+  step?: number
+  expanded?: boolean
+  disabled?: boolean
+  useHtml5Validation?: boolean
+  customClass?: string
```
Comment (daiagi) [inline libs/ui/src/components/NeoInput/NeoInput.vue:35]: I don't see customClass used

**Comment by preschian** [comment]
> test cases are failing for this part when updating to neoInput, but it seems like we dont even use this function, can we remove it? can someone confirm?

checking

**Comment by preschian** [comment]
> test cases are failing for this part when updating to neoInput

@prachi00 I think we can use `vue.extends` instead. if we want to add `.neo-*` specific class, use `vue.mixins` and then forked `rootClasess()` from `OInput`

code for `NeoInput.vue` using mixins
```vue
<script>
import { OInput } from '@oruga-ui/oruga'

// https://github.com/oruga-ui/oruga/blob/develop/packages/oruga/src/components/input/Input.vue
export default {
  mixins: [OInput],
  computed: {
    rootClasses() {
      return [
        'neo-input',
        this.computedClass('rootClass', 'o-ctrl-input'),
        {
          [this.computedClass('expandedClass', 'o-ctrl-input--expanded')]:
            this.expanded,
        },
      ]
    },
  },
}
</script>

<style lang="scss">
@import './NeoInput.scss';
</style>
```

File: libs/ui/src/components/NeoField/NeoField.vue (hunk 1 above)
[the remark this answers is not in the document]
Comment (preschian) [reply libs/ui/src/components/NeoField/NeoField.vue:11]: ref: https://github.com/kodadot/nft-gallery/pull/5870#issuecomment-1554133404

File: libs/ui/src/components/NeoInput/NeoInput.scss [hunk 6]
```diff
@@ -26,13 +25,13 @@ $input-textarea-padding: 0.5rem;
...
 
-    .o-field__label {
-      margin-bottom: 0.5rem;
-      font-weight: 400;
-      @include ktheme() {
-        color: theme('text-color');
-      }
+  .o-field__label {
+    margin-bottom: 0.5rem;
+    font-weight: 400;
```
Comment (daiagi) [inline libs/ui/src/components/NeoInput/NeoInput.scss:30]: can be replaced with:

$field-label-font-weight: 400;
$field-margin-bottom: 500;

**Review by daiagi — APPROVED** [review]
small suggestion

overall lgtm

**Comment by prachi00** [comment]
> > test cases are failing for this part when updating to neoInput
> 
> @prachi00 I think we can use `vue.extends` instead. if we want to add `.neo-*` specific class, use `vue.mixins` and then forked `rootClasess()` from `OInput`
> 
> code for `NeoInput.vue` using mixins
> 
> ```
> <script>
> import { OInput } from '@oruga-ui/oruga'
> 
> // https://github.com/oruga-ui/oruga/blob/develop/packages/oruga/src/components/input/Input.vue
> export default {
>   mixins: [OInput],
>   computed: {
>     rootClasses() {
>       return [
>         'neo-input',
>         this.computedClass('rootClass', 'o-ctrl-input'),
>         {
>           [this.computedClass('expandedClass', 'o-ctrl-input--expanded')]:
>             this.expanded,
>         },
>       ]
>     },
>   },
> }
> </script>
> 
> <style lang="scss">
> @import './NeoInput.scss';
> </style>
> ```

thank you for this, works fine now and tests are clearing as well

**Comment by VladoTheBoi** [comment]
@prachi00 checked latest commit, found just few things:
General:
- set left & right padding 16px
- remove box shadow
Dark mode:
- change border color from #CCC to #999999

**Comment by prury** [comment]
**Transfer:**
- Insert Address should begin with Capital I
- Add button needs alignment


**Identity**
- E-mail or Email instead of email
email

**Collection**
- Deployed version(Lacking some description at the bottom of it and also the char count):
symbol
- beta.kodadot.xyz:
image
---------
- deployed version: lacking char count
image
- beta.kodadot.xyz:
image


**Create NFT**
- Char counters not showing
- Field alignment doesn't match: 
image
--------
- item description when active(deployed version): 
image
- beta.kodadot.xyz:
image

**Comment by VladoTheBoi** [comment]
>Deployed version(Lacking some description at the bottom of it and also the char count):

In case there is any desciptive text lets have it like this. It also done this here: #5939 

Screenshot 2023-05-22 at 17 55 04

**Comment by prachi00** [comment]
> 

I think the margin bottom is 12px default by oruga, we can change it though, should we change to 8?

**Comment by prachi00** [comment]
Screen Shot 2023-05-23 at 7 26 03 PM
looks like this now

PS, havent made the alignment fix yet as it is using <b-select, and it would be fixed when migrating <b-select to neoselect
Screen Shot 2023-05-23 at 7 26 26 PM

**Comment by VladoTheBoi** [comment]
>I think the margin bottom is 12px default by oruga, we can change it though, should we change to 8?

Which margin do you mean exactly?

**Comment by VladoTheBoi** [comment]
> > Deployed version(Lacking some description at the bottom of it and also the char count):
> 
> In case there is any desciptive text lets have it like this. It also done this here: #5939
> 
> Screenshot 2023-05-22 at 17 55 04
> Also please change the error color to match our red which is #FF5757 (k-red) Either than that error messages can stay as is (under textfield left) until further change.

@prachi00 I posted this, not sure you've seen that. Still several things need change:
- Title is Bold
- title is 12px from textfield 
- if subtitle present, its 4px under title and 12px from textfield
- charcount is 4px under textfield, alligned right
- padding within a textfield is **16px on sides** and **12px Top&Bottom**

**Comment by vikiival** [comment]
So few things:
- @VladoTheBoi what is missing please open a follow-up issue
- @prachi00 please resolve conflicts and you have some comments from @prury, would love to merge this week

**Comment by prachi00** [comment]
> So few things:
> 
> * @VladoTheBoi what is missing please open a follow-up issue
> * @prachi00 please resolve conflicts and you have some comments from @prury, would love to merge this week

yeah we need to merge, conflicts keep popping up everyday 😢

**Comment by prachi00** [comment]
@prury I dont understand the issue with 1st video? is it that usd value is not getting updated?  I see it getting updated though so not sure

As for the 2nd one, it seems theres an icon that isnt showing when form is invalid and its taking up all the space
@roiLeo @preschian I see that the :before isnt getting applied here to icon from oruga and that is why it isnt showing up currently, would you have any idea why the icon isnt showing up?

for now, I have hidden this class so that the danger icon doesnt show up, we can solve it in another issue as this PR is getting huge 

Screen Shot 2023-05-25 at 8 11 55 PM

**Comment by preschian** [comment]
> it seems theres an icon that isnt showing when form is invalid and its taking up all the space

there is no icon named `alert-circle` on fa https://fontawesome.com/icons/alert-circle?s=solid. could you change to something else?

**Review by vikiival — COMMENTED** [review]
Code looks oki

**Comment by prury** [comment]
@prachi00 Watched my video again and its not really showing what i was trying to record, the whole site was jumping/shifting/updating when i typed a number on the amount field, but seems to be fixed now. 

The add button on the transfer page is still lacking alignment tho.

**Comment by prachi00** [comment]
> 

I'll fix the add button alignment in the followup issue with other css fixes

**Comment by prachi00** [comment]
> > it seems theres an icon that isnt showing when form is invalid and its taking up all the space
> 
> there is no icon named `alert-circle` on fa https://fontawesome.com/icons/alert-circle?s=solid. could you change to something else?


I didnt add it though, this icon was added by oruga, I'll look into replacing it if we can

**Comment by prachi00** [comment]
fixed the conflicts again, can we merge now, I'll do the rest of styling changes in followup issue 
@vikiival

**Comment by vikiival** [comment]
@prachi00 can you please open the follow-up issue with changes that needs to be done?

**Comment by prachi00** [comment]
> @prachi00 can you please open the follow-up issue with changes that needs to be done?

done

**Review by preschian — APPROVED** [review]
@prachi00 to mock matchmedia in the failed test

```js
Object.defineProperty(window, 'matchMedia', {
  writable: true,
  value: vi.fn().mockImplementation((query) => ({
    matches: false,
    media: query,
    onchange: null,
    addListener: vi.fn(),
    removeListener: vi.fn(),
    addEventListener: vi.fn(),
    removeEventListener: vi.fn(),
    dispatchEvent: vi.fn(),
  })),
})
```

> I didnt add it though, this icon was added by oruga, I'll look into replacing it if we can

yes, that one default value from oruga. need to change that in the seperate PR https://github.com/oruga-ui/oruga/blob/develop/packages/oruga/src/utils/FormElementMixin.js#L113

otherwise lgtm

**Comment by roiLeo** [comment]
> So 🥺

Screenshot 2023-05-30 at 15-13-21 #5852 Neo Input and Neo Field by prachi00 · Pull Request #5870 · kodadot_nft-gallery


knock-doorknob

**Comment by prachi00** [comment]
resolved conflicts again, can be merged now

**Comment by prachi00** [comment]
Also, e2e test cases are failing on main now, I have fixed some of them in this pr by removing it as I think it doesnt exist anymore, I hope that is correct
CC @preschian @vikiival @roiLeo

**Comment by vikiival** [comment]
>  I hope that is correct, 

So that means you don't propagate them?

**Comment by prachi00** [comment]
> > I hope that is correct,
> 
> So that means you don't propagate them?

what do you mean propagate?

**Comment by vikiival** [comment]
> what do you mean propagate?

What was the reason to remove lines from the tests.
And how to bring the back

**Comment by prachi00** [comment]
> > what do you mean propagate?
> 
> What was the reason to remove lines from the tests. And how to bring the back

because they did not exist and thats why they failed, do we even want them back?
